File: android/app/src/main/java/com/curiousbeagle/android/myaiagent/agent/AgentOutbox.kt
package com.curiousbeagle.android.myaiagent.agent

import com.curiousbeagle.android.myaiagent.model.ChatMessage
import com.curiousbeagle.android.myaiagent.net.WireMessage
import com.curiousbeagle.android.myaiagent.net.wireJson
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import java.io.File

/**
 * Per-lead session persistence: the wire conversation, the visible
 * transcript, and whether a run was in flight when last written. One JSON
 * file per lead — right-sized for a demo, and constructed with an
 * injectable directory so tests get isolated storage. Mirrors iOS
 * `AgentOutbox` exactly (the shared design deliberately uses the same
 * snapshot shape on both platforms).
 */
class AgentOutbox(private val directory: File) {

    @Serializable
    data class SessionSnapshot(
        val conversation: List<WireMessage>,
        val transcript: List<ChatMessage>,
        /**
         * True from run start until the run completes or fails. A snapshot
         * loaded with this set means the app died (or lost the network)
         * mid-run — the UI offers Resume, which replays the saved
         * conversation from the last durable boundary.
         */
        val interrupted: Boolean,
    )

    init {
        directory.mkdirs()
    }

    private fun fileFor(leadId: String) = File(directory, "$leadId.json")

    fun save(snapshot: SessionSnapshot, leadId: String) {
        runCatching { fileFor(leadId).writeText(wireJson.encodeToString(snapshot)) }
    }

    fun load(leadId: String): SessionSnapshot? = runCatching {
        wireJson.decodeFromString<SessionSnapshot>(fileFor(leadId).readText())
    }.getOrNull()

    fun clear(leadId: String) {
        fileFor(leadId).delete()
    }
}

File: iOS/MyAIAgent/MyAIAgent/Agent/AgentOutbox.swift
//
//  AgentOutbox.swift
//  MyAIAgent
//

import Foundation

/// Per-lead session persistence: the wire conversation, the visible
/// transcript, and whether a run was in flight when the app last wrote.
/// One JSON file per lead in Application Support — right-sized for a demo,
/// and initialized with a custom directory so tests get isolated storage.
nonisolated struct AgentOutbox {
    nonisolated struct SessionSnapshot: Codable, Equatable {
        var conversation: [WireMessage]
        var transcript: [ChatMessage]
        /// True from run start until the run completes or fails. A snapshot
        /// loaded with this set means the app died (or lost the network)
        /// mid-run — the UI offers Resume, which replays the saved
        /// conversation from the last durable boundary.
        var interrupted: Bool
    }

    let directory: URL

    init(directory: URL? = nil) {
        if let directory {
            self.directory = directory
        } else {
            let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            self.directory = base.appendingPathComponent("AgentSessions", isDirectory: true)
        }
        try? FileManager.default.createDirectory(at: self.directory, withIntermediateDirectories: true)
    }

    private func fileURL(for leadID: UUID) -> URL {
        directory.appendingPathComponent("\(leadID.uuidString).json")
    }

    func save(_ snapshot: SessionSnapshot, for leadID: UUID) {
        guard let data = try? JSONEncoder().encode(snapshot) else { return }
        try? data.write(to: fileURL(for: leadID), options: .atomic)
    }

    func load(for leadID: UUID) -> SessionSnapshot? {
        guard let data = try? Data(contentsOf: fileURL(for: leadID)) else { return nil }
        return try? JSONDecoder().decode(SessionSnapshot.self, from: data)
    }

    func clear(for leadID: UUID) {
        try? FileManager.default.removeItem(at: fileURL(for: leadID))
    }
}
